### ISSUE TITLE
ResultQuery.fetchOne() should not fetch second record on a LIMIT 1 query

### ISSUE THREAD

**Issue by gangeli** [jOOQ/jOOQ#7430]
### Expected behavior and actual behavior:

A `fetchOne` query such as `jooq.selectOne().fetchOne();` should, I think, have the same runtime as: `jooq.selectOne().fetchLazy().fetchNext();` I assume my observed difference is from an additional roundtrip to the server.

### Steps to reproduce the problem:

```java
long startTime = System.currentTimeMillis();
try (DSLContext jooq = DSL.using(conn)) {
  jooq.selectOne().fetchLazy().fetchNext();
  // or: jooq.selectOne().fetchOne();
} finally {
  System.out.println("Query took: " + (System.currentTimeMillis() - startTime) + "ms");
}
```

After some burn-in, for `#fetchOne` I get:

```
17:00:30.023 [LoggerListener]               DEBUG: Executing query          : select 1 as `one` from dual
17:00:30.125 [LoggerListener]               TRACE: Record fetched           : +----+
17:00:30.125 [LoggerListener]               TRACE:                          : | one|
17:00:30.125 [LoggerListener]               TRACE:                          : +----+
17:00:30.126 [LoggerListener]               TRACE:                          : |   1|
17:00:30.126 [LoggerListener]               TRACE:                          : +----+
17:00:30.126 [LoggerListener]               DEBUG: Fetched result           : +----+
17:00:30.126 [LoggerListener]               DEBUG:                          : | one|
17:00:30.126 [LoggerListener]               DEBUG:                          : +----+
17:00:30.126 [LoggerListener]               DEBUG:                          : |   1|
17:00:30.127 [LoggerListener]               DEBUG:                          : +----+
"Query took 152ms"  // 2x50ms roundtrip? + 50ms disk access
```

Whereas, for `#fetchLazy().fetchNext()`, I get:

```
16:52:48.141 [LoggerListener]               DEBUG: Executing query          : select 1 as `one` from dual
16:52:48.243 [LoggerListener]               TRACE: Record fetched           : +----+
16:52:48.243 [LoggerListener]               TRACE:                          : | one|
16:52:48.243 [LoggerListener]               TRACE:                          : +----+
16:52:48.243 [LoggerListener]               TRACE:                          : |   1|
16:52:48.243 [LoggerListener]               TRACE:                          : +----+
16:52:48.244 [LoggerListener]               DEBUG: Fetched result           : +----+
16:52:48.245 [LoggerListener]               DEBUG:                          : | one|
16:52:48.245 [LoggerListener]               DEBUG:                          : +----+
16:52:48.245 [LoggerListener]               DEBUG:                          : |   1|
16:52:48.245 [LoggerListener]               DEBUG:                          : +----+
"Query took 103ms"  // 1x50ms roundtrip? + 50ms disk access
```

My ping is:

```
$ ping <redacted>
PING <redacted> (<redacted>) 56(84) bytes of data.
64 bytes from mysql (<redacted>): icmp_seq=1 ttl=40 time=50.5 ms
64 bytes from mysql (<redacted>): icmp_seq=2 ttl=40 time=50.4 ms
64 bytes from mysql (<redacted>): icmp_seq=3 ttl=40 time=50.5 ms
64 bytes from mysql (<redacted>): icmp_seq=4 ttl=40 time=50.6 ms
64 bytes from mysql (<redacted>): icmp_seq=5 ttl=40 time=50.5 ms
...
```

### Versions:

- jOOQ: 3.10.5
- Java: 1.8
- Database (include vendor): MySQL (Oracle)
- OS: Linux (Ubuntu 14.04 x64)
- JDBC Driver (include name if inofficial driver): mysql-connector-java:8.0.11

**Comment by lukaseder** [comment]
Thanks for your report. There are several things to note here:

1. You turned on trace logging. This incurs its own overhead, i.e. time spent in the logger, which may affect the perceived execution time. The penalty is "fair", because both of your measurements have the penalty, but it's worth mentioning that the numbers may be skewed for a variety of reasons, including this one.
2. `fetchOne()` throws a `TooManyRowsException` in case the database returns more than one row. In order to *know* whether this exception needs to be thrown, the second record has to be fetched as well as the first one. If you don't care about that, you might prefer using `fetchAny()` (which does `fetchLazy().fetchNext()`, internally)
3. Perhaps you could tamper with the JDBC `Statement.setFetchSize()` flag (or jOOQ's `ResultQuery.fetchSize()`) to fetch all the rows in one go, in case you're streaming them.
4. Be careful with `jooq.selectOne().fetchLazy().fetchNext();`. `fetchLazy()` returns `org.jooq.Cursor`, which is a resource (containing the JDBC `ResultSet`), and should be closed eagerly using try-with-resources
5. 50ms per roundtrip does seem excessive. Is this a realistic production-like scenario in your case?

I hope this helps

**Comment by gangeli** [comment]
Thanks Lukas!

The second call to decide whether we need to throw an exception or not makes a lot of sense. I'll use `#fetchAny` in the future when I'm sure there's only one return (e.g., I "LIMIT 1" in the query).

And don't worry, our production environment has a reasonable ping -- this is just running from my local machine. :)

**Comment by lukaseder** [comment]
Huh, if jOOQ *knows* there's a `LIMIT 1` clause, perhaps we could avoid fetching the second row, because there can be none (except when `WITH TIES` semantics is applied).

I will change this issue to fix that.

**Comment by lukaseder** [comment]
Version 3.10.6 looks quite different from 3.11, due to #7001. We thus won't backport this fix to 3.10 or 3.9

**Comment by lukaseder** [comment]
Thanks again for reporting this!
